uniswap_edpr/src/logics/shared_algorithm_lens/src/lib.rs

```diff
@@ -0,0 +1,86 @@
+  let pool_fees_result = get_get_last_snapshot_value_in_pool_fees (targets . get (0usize) . unwrap () . clone ()) . await ;
+  let tc28x6_result = get_get_last_snapshot_value_in_tcumul_28x6hr (targets . get (1usize) . unwrap () . clone ()) . await ;
+  let v3pool_result = get_get_last_snapshot_value_in_v3pool (targets . get (2usize) . unwrap () . clone ()) . await ;
+  let eth_usdc_price_result = get_get_last_snapshot_value_in_eth_usdc_price (targets . get (3usize) . unwrap () . clone ()) . await ;
+  
+  let fees_24h_usd = pool_fees_result.clone().unwrap().result[0].fees_24h_usd;
+  let t0_decimals = pool_fees_result.clone().unwrap().result[0].pool_summary_level_1.t0_decimals;
+  let t1_decimals =  pool_fees_result.clone().unwrap().result[0].pool_summary_level_1.t1_decimals;
+
+  let eth_price_usd  = eth_usdc_price_result.clone().unwrap().result.t1_price_usd;
+
+  let address = v3pool_result.clone().unwrap().result.address;
+  let current_tick_liquidity_str = v3pool_result.clone().unwrap().result.liquidity;
+  let current_tick_liquidity = i128::from_str_radix(current_tick_liquidity_str.trim_start_matches("0x"), 16).unwrap();
+  let sqrt_ratio_x96_str = v3pool_result.clone().unwrap().result.sqrt_ratio_x96;
+  let sqrt_ratio_x96 = u128::from_str_radix(sqrt_ratio_x96_str.trim_start_matches("0x"), 16).unwrap();
+  let tick_current = v3pool_result.clone().unwrap().result.tick_current;
+  let tick_spacing = v3pool_result.clone().unwrap().result.tick_spacing;
+  let token0 = v3pool_result.clone().unwrap().result.token0;
+
+  let tick_cumul_28x6h = tc28x6_result.unwrap().0;
+
+  let current_sqrt_price = sqrt_ratio_x96 as f32 / f32::powf(2.0,96.0);
+  let mut current_price = current_sqrt_price * current_sqrt_price;
+
+  let mut compressed: i32 = tick_current / tick_spacing;
+  if tick_current <0 && tick_current % tick_spacing != 0 {
+    compressed -= 1;
+  }
+  let floor = compressed * tick_spacing;
+  let mut min_tick = floor;
+  let mut max_tick = floor + tick_spacing;
+  for i in (1..29).rev() {
+    let new_tick = ((tick_cumul_28x6h[i] - tick_cumul_28x6h[i-1]) / 21600) as i32;
+    if new_tick < min_tick {
+      min_tick = new_tick;
+    } else if new_tick > max_tick {
+      max_tick = new_tick;
+    }
+  }
+  
+  let mut range_top = f32::powf(1.0001, max_tick as f32);
+  let mut range_bottom = f32::powf(1.0001, min_tick as f32);
+  let amount0 = current_tick_liquidity as f32
+    * (range_top.sqrt() - current_sqrt_price as f32)
+    / (current_sqrt_price as f32 * range_top.sqrt())
+    / f32::powf(10.0, t0_decimals as f32);
+  let amount1 = current_tick_liquidity as f32
+    * (current_sqrt_price as f32 - range_bottom.sqrt())
+    / f32::powf(10.0, t1_decimals as f32);
+  let est_tvl_in_range = amount0 * current_price + amount1;
+
+  if token0 == "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2" {
+    current_price = 1.0 / current_price;
+    let est_tvl_in_range = amount0 + amount1 * current_price;
```

Comment:
This does not appear to be used.
Do I need to override est_tvl_in_range one scope up?

Reply:
Yes, you do. My oversight, if I've understand your question correctly. Should I therefore use `let mut ...` in line 63?

Reply:
I tested this out in the Rust Playground.

Current version (modified so it works in the Playground) - https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=4258daae15e50208d13786547aec43f2

Output:
```
   Compiling playground v0.0.1 (/playground)
warning: unused variable: `est_tvl_in_range`
  --> src/main.rs:71:9
   |
71 |     let est_tvl_in_range = amount0 + amount1 * current_price;
   |         ^^^^^^^^^^^^^^^^ help: if this is intentional, prefix it with an underscore: `_est_tvl_in_range`
   |
   = note: `#[warn(unused_variables)]` on by default

warning: `playground` (bin "playground") generated 1 warning (run `cargo fix --bin "playground"` to apply 1 suggestion)
    Finished dev [unoptimized + debuginfo] target(s) in 1.01s
     Running `target/debug/playground`
Standard Output
LensValue { address: "0xe42318ea3b998e8355a3da364eb9d48ec725eb45", current_price: 0.013549291, range_top: 0.014449563, range_bottom: 0.012134394, edpr: 0.0005491668 }
```

Added `let mut` to line 67 and removed `let` from line 71 - https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=51f027b706edf9ec0791c611c8e5be9b

Output:
```
   Compiling playground v0.0.1 (/playground)
    Finished dev [unoptimized + debuginfo] target(s) in 0.78s
     Running `target/debug/playground`
Standard Output
LensValue { address: "0xe42318ea3b998e8355a3da364eb9d48ec725eb45", current_price: 0.013549291, range_top: 0.014449563, range_bottom: 0.012134394, edpr: 0.04053104 }
```

Here's a manual calculation of EDPR using current data from https://info.uniswap.org/#/pools/0xe42318ea3b998e8355a3da364eb9d48ec725eb45 :

The selected range corresponds to 29 tick intervals. 
ETH locked within the current tick interval = 43.45 ETH ≡ 43.45 * 2520 USD = 109494 USD
The estimated TVL in range = 29 * 109494 = 3175326 USD
24h Fees = 4690 USD
Estimated daily return = 4690 / 3175326 = 0.00147701369 ≈ 0.148%

As I noted somewhere in our other discussion, info.uniswap has inherent inaccuracies and sometimes [greatly over- or under-estimates TVL](https://medium.com/coinmonks/all-your-uniswap-v3-liquidity-farming-calculations-are-dead-wrong-heres-why-20bd47f55d69), so I think the second edpr result looks much more realistic than the first result.

[Modifying the Playground code](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=28170277cb356df83f8e4fabf09c743d) to display `est_tvl_in_range` gives a result of 4077.8281 ETH ≡ 4077.8281 * 2374.3784 USD ≈ 9682307 USD.

Using the same figures as in the code, estimated daily return = 3924.3396 / 9682307 ≈ 0.00040531 ≡ 0.040531, which matches the result obtained by the revised code.

Thank you for spotting this! I'll modify the code accordingly.